src/posix/platform/netif.cpp

```diff
@@ -379,6 +793,284 @@ static void processNetifLinkEvent(otInstance *aInstance, struct nlmsghdr *aNetli
+#define RT_ROUNDUP(a) RT_ROUNDUP2((a), sizeof(uint64_t))
+#define SA_SIZE(sa) RT_ROUNDUP(sa->sa_len)
+#endif
+
+static void processNetifAddrEvent(otInstance *aInstance, struct rt_msghdr *rtm) //, size_t buffer_len)
+{
+    otError            error;
+    struct ifa_msghdr *ifam;
+#ifdef RTM_NEWMADDR
+    struct ifma_msghdr *ifmam;
+#endif
+    struct sockaddr_in6 addr6;
+    struct sockaddr_in6 netmask;
+    uint8_t *           addrbuf;
+    unsigned int        addrlen;
+    unsigned int        addrmask = 0;
+    unsigned int        i;
+    struct sockaddr *   sa;
+    bool                is_link_local;
+    char                addressString[INET6_ADDRSTRLEN + 1];
+    size_t              buffer_len = rtm->rtm_msglen;
+
+    addr6.sin6_family   = 0;
+    netmask.sin6_family = 0;
+
+    if ((rtm->rtm_type == RTM_NEWADDR) || (rtm->rtm_type == RTM_DELADDR))
+    {
+        ifam = reinterpret_cast<struct ifa_msghdr *>(rtm);
+
+        VerifyOrExit(ifam->ifam_index == static_cast<unsigned int>(sTunIndex), OT_NOOP);
+
+        addrbuf  = (uint8_t *)&ifam[1];
+        addrmask = (unsigned int)ifam->ifam_addrs;
+    }
+#ifdef RTM_NEWMADDR
+    else if ((rtm->rtm_type == RTM_NEWMADDR) || (rtm->rtm_type == RTM_DELMADDR))
+    {
+        ifmam = reinterpret_cast<struct ifma_msghdr *>(rtm);
+
+        VerifyOrExit(ifmam->ifmam_index == static_cast<unsigned int>(sTunIndex), OT_NOOP);
+
+        addrbuf  = (uint8_t *)&ifmam[1];
+        addrmask = (unsigned int)ifmam->ifmam_addrs;
+    }
+#endif
+    addrlen = (unsigned int)buffer_len;
+
+    if (addrmask != 0)
+    {
+        for (i = 0; i < RTAX_MAX; i++)
+        {
+            unsigned int mask = (addrmask & (1 << i));
+            if (mask)
+            {
+                sa = (struct sockaddr *)addrbuf;
+
+                if (sa->sa_family == AF_INET6)
+                {
+                    if (i == RTAX_IFA)
+                        memcpy(&addr6, sa, sizeof(sockaddr_in6));
+                    if (i == RTAX_NETMASK)
+                        memcpy(&netmask, sa, sizeof(sockaddr_in6));
+                }
+                addrlen -= SA_SIZE(sa);
+                addrbuf += SA_SIZE(sa);
+            }
+        }
+    }
+
+    if (addr6.sin6_family == AF_INET6)
+    {
+        is_link_local = false;
+        if (IN6_IS_ADDR_LINKLOCAL(&addr6.sin6_addr))
+        {
+            is_link_local = true;
+            // clear the scope -- Mac OS X sends this to us (bozos!)
+            addr6.sin6_addr.s6_addr[3] = 0;
+        }
+        else if (IN6_IS_ADDR_MC_LINKLOCAL(&addr6.sin6_addr))
+        {
+            addr6.sin6_addr.s6_addr[3] = 0;
+        }
+
+        ot::Ip6::Address addr;
+        memcpy(&addr, &addr6.sin6_addr, sizeof(addr));
+
+        if (rtm->rtm_type == RTM_NEWADDR
+#ifdef RTM_NEWMADDR
+            || rtm->rtm_type == RTM_NEWMADDR
+#endif
+        )
+        {
+            if (!addr.IsMulticast())
+            {
+                otNetifAddress netAddr;
```

Comment:
Move this long code block to a separate function?

Reply:
I consolidated a lot of the logging that was making this so complicated.  Take a peek.